### PR TITLE
Storage

### PR DESCRIPTION
**Pull request by philipstanislaus**

### CODE REVIEW

File: scale/codec.go [hunk 1]
```diff
@@ -197,20 +208,12 @@ func (pe Encoder) Encode(value interface{}) error {
 		}
 
 	case reflect.Struct:
-		encodeable := reflect.TypeOf((*Encodeable)(nil)).Elem()
-		if t.Implements(encodeable) {
```
Comment (vimukthi-git) [inline scale/codec.go]: I wonder why this existing code doesn't work for you ?

File: scale/codec.go [hunk 2]
```diff
@@ -408,22 +424,11 @@ func (pd Decoder) DecodeIntoReflectValue(target reflect.Value) error {
 		target.SetString(string(b))
 
 	case reflect.Struct:
-		encodeable := reflect.TypeOf((*Decodeable)(nil)).Elem()
```
Comment (vimukthi-git) [inline scale/codec.go]: same here

File: scale/codec.go (hunk 1 above)
Comment (philipstanislaus) [reply scale/codec.go]: It did work, I just moved the check for encodeable to the top. Before, you could not specify custom encoding for non-structs.

File: scale/codec.go (hunk 2 above)
Comment (philipstanislaus) [reply scale/codec.go]: See above

File: types/storage_key.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+			return nil, err
+		}
+	} else if fnMeta.isDMap() {
+		// TODO define hashing for 2 keys
+		return nil, fmt.Errorf("double map storage keys are not yet implemented")
+	}
+
+	afn := []byte(module + " " + fn)
+
+	// TODO why is add length prefix step in JS client doesn't add anything to the hashed key?
```
Comment (vimukthi-git) [inline types/storage_key.go]: we can remove this `TODO` as its no longer valid. The reason is that the hash length is known upfront by the encoders and decoders.

Comment (philipstanislaus) [reply types/storage_key.go]: Done

**Pull request merged**